### ISSUE TITLE
PENDING_TX_DOWNLOADS_LIMIT is somewhat arbitrary and too low in at least some use-cases

### ISSUE THREAD

**Issue by pjakma** [bitcoinj/bitcoinj#3210]
The code in Peer::blockChainDownloadLocked contains a check of the number of pending transactions against a hard-coded limit, PENDING_TX_DOWNLOADS_LIMIT. If the limit is exceeded, the connection is simply disconnected. See:

https://github.com/bitcoinj/bitcoinj/blob/19fcdcdd5ba3d215b1a9999881ebe046a98c218c/core/src/main/java/org/bitcoinj/core/Peer.java#L1191C49-L1191C75

This limit seems somewhat arbitrary, and it is getting tripped regularly in an app that uses Bitcoinj that I use regularly. Causing problematic connection disconnects, and reliability issues.

Could this limited be increased? Ideally, many fold (500+). Or could it be made configurable/overrideable to the using app via API in some way?

**Comment by schildbach** [comment]
Of course it would be easy to bump the limit. However the real question is why is the limit tripped? The `pendingTxDownloads` set only contains entries that have been requested (`GETDATA`) from the peer. Usually it should respond quickly by sending the transaction and the entry is removed from the set. In general, the set should not grow.

Thinking about the possible edge cases, I see one potential reason for leakage of   `pendingTxDownloads` entries:

If a block confirms some pending txns contained in the set, we can't `GETDATA` them any more. Such entries would stay there forever. Maybe we should be actively removing them if a block is processed?

**Comment by schildbach** [comment]
Can you try to reproduce your issue against a bitcoind that is under your control? Once reproduced, could you capture the logs of both bitcoinj and bitcoind for one session? We should be able to tell from the logs if something goes wrong.

**Comment by pjakma** [comment]
@schildbach Yes, this is happening with a connection to a local bitcoin. I'll try get the logs on both sides.
